google-cloud-spanner/src/test/java/com/google/cloud/spanner/it/ITDmlReturningTest.java

```diff
@@ -0,0 +1,263 @@
+  }
+
+  private void executeUpdate(long expectedCount, final String... stmts) {
+    final TransactionCallable<Long> callable =
+        transaction -> {
+          long rowCount = 0;
+          for (String stmt : stmts) {
+            rowCount += transaction.executeUpdate(Statement.of(stmt));
+          }
+          return rowCount;
+        };
+    TransactionRunner runner = getClient(dialect.dialect).readWriteTransaction();
+    Long rowCount = runner.run(callable);
+    assertThat(rowCount).isEqualTo(expectedCount);
+  }
+
+  @Test
+  public void dmlReturningWithExecutePartitionedUpdate() {
+    Throwable exception =
+        assertThrows(
+            SpannerException.class,
+            () ->
+                getClient(dialect.dialect)
+                    .executePartitionedUpdate(Statement.of(updateDmlReturning())));
+    assertThat(exception.getMessage()).ignoringCase().contains("not supported");
+  }
+
+  @Test
+  public void dmlReturningWithExecuteQuery() {
+    List<Struct> rows = executeQuery(DML_COUNT, insertDmlReturning());
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V"))
+                .getLong(0))
+        .isEqualTo(1);
+
+    // executeQuery results have keys(K) sorted.
+    // Check if keys(K) and V have expected values.
+    for (int idx = 0; idx < rows.size(); idx++) {
+      assertThat(rows.get(idx).getLong("V")).isEqualTo(idx + 1);
+      assertThat(rows.get(idx).getString("K")).isEqualTo(String.format("%d-boo%d", id, idx + 1));
+    }
+    rows = executeQuery(DML_COUNT, updateDmlReturning());
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V"))
+                .getLong(0))
+        .isEqualTo(100);
+
+    // executeQuery results have keys(K) sorted.
+    // Check if keys(K) and V have expected values.
+    for (int idx = 0; idx < rows.size(); idx++) {
+      assertThat(rows.get(idx).getLong("V")).isEqualTo(100);
+      assertThat(rows.get(idx).getString("K")).isEqualTo(String.format("%d-boo%d", id, idx + 1));
+    }
+    rows = executeQuery(DML_COUNT, deleteDmlReturning());
+    assertThat(
+            getClient(dialect.dialect)
+                .singleUse(TimestampBound.strong())
+                .readRow("T", Key.of(String.format("%d-boo1", id)), Collections.singletonList("V")))
+        .isNull();
+
+    // executeQuery results have keys(K) sorted.
+    // Check if keys(K) and V have expected values.
+    for (int idx = 0; idx < rows.size(); idx++) {
+      assertThat(rows.get(idx).getLong("V")).isEqualTo(100);
+      assertThat(rows.get(idx).getString("K")).isEqualTo(String.format("%d-boo%d", id, idx + 1));
+    }
+  }
+
+  private List<Struct> executeQuery(long expectedCount, String stmt) {
```

Comment:
Can you add regular DML testing over this read ExecuteQuery API?

Reply:
Regular DML testing over executeQuery API is already available at https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/test/java/com/google/cloud/spanner/it/ITDMLTest.java#L384. So, I didn't add it here to prevent redundant tests.